DRAFT prospective changes from WGLC review (#171)
oauth-wg/oauth-identity-chaining: draft-ietf-oauth-identity-chaining.md

This specification describes a combination of OAuth 2.0 Token Exchange {{RFC8693}} and JWT Profile for OAuth 2.0 Client Authentication and Authorization Grants {{RFC7523}} to achieve identity and authorization chaining across domains.

A client in trust domain A that needs to access a resource server in trust domain B requests a JWT authorization grant from the authorization server for trust domain A using a profile of OAuth 2.0 Token Exchange {{RFC8693}}. The client in trust domain A then presents the received grant as an assertion to the authorization server in trust domain B, using the JWT Authorization Grant feature of {{RFC7523}}, to obtain an access token for the protected resource in trust domain B.

> In some deployments, the platform in trust domain A may be sufficiently integrated to directly produce a JWT authorization grant for presentation to the authorization server in trust domain B. In such cases, use of the OAuth 2.0 Token Exchange protocol {{RFC8693}} is not required, as the JWT authorization grant can be obtained without a separate token exchange step.

Reviewer: Instead of "platform" how about just saying "the authorization server"?

> In some deployments, the platform in trust domain A may be sufficiently integrated to directly produce a JWT authorization grant for presentation to the authorization server in trust domain B. In such cases, use of the OAuth 2.0 Token Exchange protocol {{RFC8693}} is not required, as the JWT authorization grant can be obtained without a separate token exchange step.

Author: The kind of situation I was trying to account for is like what's described in the unfortunately named but conceptually interesting Github OIDC for actions thing https://docs.github.com/en/actions/concepts/security/openid-connect where the Github platform has integrated APIs (like, again the unfortunately named getIDToken() methods) of getting the grant directly from the platform using the runtime context or session. 

It doesn't seem like my words have captured this very well. But I'm not sure the suggestions you gentlemen have offered do either. Maybe.

Thoughts?

Different suggestions after that attempt at explaining myself?

> In some deployments, the platform in trust domain A may be sufficiently integrated to directly produce a JWT authorization grant for presentation to the authorization server in trust domain B. In such cases, use of the OAuth 2.0 Token Exchange protocol {{RFC8693}} is not required, as the JWT authorization grant can be obtained without a separate token exchange step.

Author: Yes, I think that is what I'm trying to describe here. I don't want to try to spec it out or even describe it in detail but did want to have a brief mention of the possibility in the hopes that people don't think that kind of deployment is precluded by this and go off and build something else like github did. Not that a sentence or two will actually prevent that kind of thing but I wanted to at least try to have something.

> In some deployments, the platform in trust domain A may be sufficiently integrated to directly produce a JWT authorization grant for presentation to the authorization server in trust domain B. In such cases, use of the OAuth 2.0 Token Exchange protocol {{RFC8693}} is not required, as the JWT authorization grant can be obtained without a separate token exchange step.

Reviewer 2: To me it seems like we want to open up the possibility to replace token exchange with something else when it's more convenient, is that correct?

> In some deployments, the platform in trust domain A may be sufficiently integrated to directly produce a JWT authorization grant for presentation to the authorization server in trust domain B. In such cases, use of the OAuth 2.0 Token Exchange protocol {{RFC8693}} is not required, as the JWT authorization grant can be obtained without a separate token exchange step.

Reviewer: I am a bit unsure about calling it out - I think it is a pattern people use, but it raises the question on whether this will encourage more people to do somehting non standard and then claim "I am am implementing the standard".

I tried to write sown what that might look like, which also feels... odd... basically saying, use the protocol,. unless your not using the protocol.

```suggestion
In some deployments, the client in trust domain A may obtain a JWT authorization grant issued by the  authorization in trust domain A using a proprietary API or interface other than the OAuth 2.0 Token Exchange protocol {{RFC8693}}.

> In some deployments, the platform in trust domain A may be sufficiently integrated to directly produce a JWT authorization grant for presentation to the authorization server in trust domain B. In such cases, use of the OAuth 2.0 Token Exchange protocol {{RFC8693}} is not required, as the JWT authorization grant can be obtained without a separate token exchange step.

Author: > I am a bit unsure about calling it out - I think it is a pattern people use, but it raises the question on whether this will encourage more people to do somehting non standard and then claim "I am am implementing the standard".
> 
> I tried to write sown what that might look like, which also feels... odd... basically saying, use the protocol,. unless your not using the protocol.

Yeah, I get that. But there are two parts of this protocol that are nice together but not required together. And I'd argue that the second part, the x-domain part, is the more important from a interoperability and health of the broader industry. And I'm tying to avoid people doing something non-standard there (as unfortunately has been done) because they misinterpret this doc as precluding them because they have a "more convenient" way to get the grant thing.

> In some deployments, the platform in trust domain A may be sufficiently integrated to directly produce a JWT authorization grant for presentation to the authorization server in trust domain B. In such cases, use of the OAuth 2.0 Token Exchange protocol {{RFC8693}} is not required, as the JWT authorization grant can be obtained without a separate token exchange step.

Author: or with more words something like this:
```suggestion
In some deployments, the client in trust domain A may obtain a JWT authorization grant using a proprietary API or interface other than the OAuth 2.0 Token Exchange protocol {{RFC8693}}. The details of such an interface are out of scope for this document but an alternative means of acquiring the JWT authorization grant is not precluded by this document and its other aspects are still useful.
```

Authorization servers MAY transcribe claims when either producing JWT authorization grants in the token exchange flow or access tokens in the assertion flow. Transcription of claims may be required for the following reasons:

* **Transcribing the subject identifier**: The subject identifier can differ between the parties involved. For example, a user is identified in trust domain A as "johndoe@a.org" but in trust domain B they are identified as "doe.john@b.org". The mapping from one identifier to the other MAY either happen in the token exchange step and the updated identifier is reflected in the returned JWT authorization grant or in the assertion step where the updated identifier would be reflected in the access token. To support this both authorization servers MAY add, change or remove claims as described above.
* **Data Minimization**: Authorization servers MAY remove or hide certain claims due to privacy requirements or reduced trust towards the targeting trust domain.
One example is a financial institution that integrates with a third-party payment gateway.

> domain A (the financial institution) includes detailed claims such as "account type: premium" and "transaction limit: $10,000" in the JWT authorization grant.

Reviewer: It's abit confusing with the tooling, bu it looks to me like this is the start of a new sentence, so perhaps we should keep Domain capitalised in this case?

> domain A (the financial institution) includes detailed claims such as "account type: premium" and "transaction limit: $10,000" in the JWT authorization grant.

Author: ```suggestion
Domain A (the financial institution) includes detailed claims such as "account type: premium" and "transaction limit: $10,000" in the JWT authorization grant.
```
